cxx-qt-gen: (dis)connect from signals (#485)
KDAB/cxx-qt: book/src/qobject/signals_enum.md

Note that `#[cxx_name = "..."]` can also be used on a signal to declare a different name in C++ to Rust.

## Connecting to a signal

> For every signal defined in the enum a method is generated `on_<signal_name>`,

Reviewer: ```suggestion
For every signal defined in the enum, a method is generated called `on_<signal_name>`.
```

Author: Done

> this method takes a function pointer as the fist parameter with the first argument being the qobject
> and the remaining arguments being the signal parameters, the second parameter is the Qt connection type.

Reviewer: ```suggestion
This method takes a function as the first parameter. That function's first argument is the qobject
and the remaining arguments are the signal parameters. The second parameter of the `on_<signal_name>` method is the [Qt connection type](https://doc.qt.io/qt-6/qt.html#ConnectionType-enum).
```

Author: Done

Note that by using the `#[inherit]` macro on a signal, connections can be made to property changes
using the signal name `<property>Changed` with no parameters.

```rust,ignore,noplayground
{{#include ../../../examples/qml_features/rust/src/signals.rs:book_signals_connect}}
```

> Each connection returns a `QMetaObject::Connection` which then has an extra `disconnect` method
> in Rust which allows you to disconnect the Rust function pointer from the signal.

Reviewer: ```suggestion
Each connection returns a `QMetaObject::Connection` which can be disconnected later by calling its `disconnect` method.
```

Author: Done

> Each connection returns a [`QMetaObject::Connection`](https://doc.qt.io/qt-6/qmetaobject-connection.html) which can be disconnected later by calling its `disconnect` method.

Author: need to update this doc

Author: Updated.

Note that `#[cxx_name = "..."]` can also be used on a signal to declare a different name in C++ to Rust.

## Connecting to a signal

> For every signal defined in the enum, a method is generated called `on_<signal_name>`.
> This method takes a function as the first parameter.
> That function's first argument is the qobject and the remaining arguments are the signal parameters.
> The second parameter of the `on_<signal_name>` method is the [Qt connection type](https://doc.qt.io/qt-6/qt.html#ConnectionType-enum).

Reviewer 2: ```suggestion
For every signal defined in the enum, two methods are generated.

1. `on_<signal_name>`
2. `connect_<signal_name>`

The `on_<signal_name>` method takes a handler function as the parameter, which will be called when the signal is emitted.
That handler function's first argument is the qobject and the remaining arguments are the signal parameters.

The `connect_<signal_name>` function additionally takes the [Qt connection type](https://doc.qt.io/qt-6/qt.html#ConnectionType-enum) as a parameter.
```

Author: done

Note that by using the `#[inherit]` macro on a signal, connections can be made to property changes
using the signal name `<property>Changed` with no parameters.

```rust,ignore,noplayground
{{#include ../../../examples/qml_features/rust/src/signals.rs:book_signals_connect}}
```

> Each connection returns a [`QMetaObject::Connection`](https://doc.qt.io/qt-6/qmetaobject-connection.html) which can be disconnected later by `drop` occurring on the type,
> if you don't want a disconnect to occur call `release`.

Reviewer 2: ```suggestion
Each connection returns a [`QMetaObject::Connection`](https://doc.qt.io/qt-6/qmetaobject-connection.html) which is used to manage the connection.
Note that the `QMetaObjectConnection` returned by Cxx-Qt behaves a bit different from the Qt C++ implementation.
When the `QMetaObjectConnection` is dropped, it automatically disconnects the connection, similar to how a C++ `std::unique_ptr` or Rusts `Box` behaves.
If you don't want to store the QMetaObjectConnection, call `release`, which will drop the object without disconnectiong.
In this case, it is no longer possibly to disconnect later.
```

Author: Done
